packages/admin/cms-admin/src/common/MasterMenu.tsx

```diff
@@ -13,68 +16,122 @@ import { RouteProps, useRouteMatch } from "react-router-dom";
 
 import { useUserPermissionCheck } from "../userPermissions/hooks/currentUser";
 
-type MasterMenuItemRoute = Omit<MenuItemRouterLinkProps, "to"> & {
+type MasterMenuItemBase = {
     requiredPermission?: string;
-    route?: RouteProps;
-    to?: string;
-    submenu?: MasterMenuItem[];
 };
 
-type MasterMenuItemAnchor = MenuItemAnchorLinkProps & {
-    requiredPermission?: string;
-};
+type MasterMenuItemRoute = MasterMenuItemBase &
+    Omit<MenuItemRouterLinkProps, "to"> & {
+        type: "route";
+        route?: RouteProps;
+        to?: string;
+    };
 
-export type MasterMenuItem = MasterMenuItemRoute | MasterMenuItemAnchor;
+type MasterMenuItemCollapsible = MasterMenuItemBase &
+    Omit<MenuCollapsibleItemProps, "children"> & {
+        type: "collapsible";
+        submenu?: Array<MasterMenuItemRoute | MasterMenuItemAnchor | MasterMenuItemCollapsible>;
+        route?: RouteProps;
+    };
+
+type MasterMenuItemAnchor = MasterMenuItemBase &
+    MenuItemAnchorLinkProps & {
+        type: "anchor";
+    };
+
+type MasterMenuItemGroup = MasterMenuItemBase &
+    MenuItemGroupProps & {
+        type: "group";
+        groupItems: Array<MasterMenuItemRoute | MasterMenuItemAnchor | MasterMenuItemCollapsible>;
+    };
+
+export type MasterMenuItem = MasterMenuItemRoute | MasterMenuItemAnchor | MasterMenuItemCollapsible | MasterMenuItemGroup;
 
 export type MasterMenuData = MasterMenuItem[];
 
-export function isMasterMenuItemAnchor(item: MasterMenuItem): item is MasterMenuItemAnchor {
-    return !!item.href;
+type MenuItemRouteElement = {
+    type: "route";
+    menuElement: MenuItemRouterLinkProps;
+};
+
+type MenuCollapsibleElement = {
+    type: "collapsible";
+    menuElement: Omit<MenuCollapsibleItemProps, "children">;
+    submenu: MenuItem[];
+};
+
+type MenuItemAnchorElement = {
+    type: "anchor";
+    menuElement: MenuItemAnchorLinkProps;
+};
+
+type MenuItemGroupElement = {
+    type: "group";
+    menuElement: MenuItemGroupProps;
+    groupItems: MenuItem[];
+};
+
+type MenuItem = MenuItemRouteElement | MenuItemAnchorElement | MenuCollapsibleElement | MenuItemGroupElement;
+
+export interface MasterMenuProps {
+    permanentMenuMinWidth?: number;
+    menu: MasterMenuData;
+}
+
+export function isMenuItemGroup(item: MenuItem): item is MenuItemGroupElement {
+    return item.type === "group";
+}
+export function isMenuItemAnchor(item: MenuItem): item is MenuItemAnchorElement {
+    return item.type === "anchor";
+}
+
+export function isMenuItemRoute(item: MenuItem): item is MenuItemRouteElement {
+    return item.type === "route";
+}
+
+export function isMenuItemCollapsible(item: MenuItem): item is MenuCollapsibleElement {
+    return item.type === "collapsible";
 }
```

Comment:
Do we even need this type guards? Shouldn't checking on `type` be sufficient?

Reply:
Removed in 98869e6328d14704a437f3f6300dacd8d0fb9ac6